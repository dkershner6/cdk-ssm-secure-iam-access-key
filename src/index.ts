import * as fs from 'fs';
import * as path from 'path';

import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as lambdanodejs from 'aws-cdk-lib/aws-lambda-nodejs';
import * as cr from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';

export { IAMAccessKey } from './provider';

export interface ISSMSecureIAMAccessKeyProps {
  readonly parameterName: string;
  readonly region?: string;
  readonly user: iam.IUser;
}

/**
 * Custom resource that creates a new IAM access key for the given user, then stores the AccessKeyId and Secret in a SSM SecureString Parameter.
 *
 * @param props SSMSecureIAMAccessKeyProps
 *
 * @returns (Not actually returned) The Value of the Parameter is the JSON representation of the AccessKeyId and Secret:
 * @see IAMAccessKey
 * Example: { "accessKeyId": "AKIAXXXXXXXXXXXXXXXX", "secretAccessKey": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX" }
 *
 * Give permission to use this secret SSM Parameter:
 * ssm.StringParameter.fromSecureStringParameterAttributes(this, "ID", {
 *     parameterName: this.props.parameterName,
 * });
 */
export class SSMSecureIAMAccessKey extends Construct {
  private readonly provider: cr.Provider;

  constructor(
    scope: Construct,
    private readonly id: string,
    private readonly props: ISSMSecureIAMAccessKeyProps,
  ) {
    super(scope, id);

    this.validateProps();

    this.provider = this.createAccessKeyCreatorProvider();
    this.createResource();
  }

  private validateProps(): void {
    if (!this.props.parameterName) {
      throw new Error('parameterName is required');
    }
    if (!this.props.user?.userName) {
      throw new Error('user is required');
    }
  }

  private createAccessKeyCreatorProvider(): cr.Provider {
    const providerFileName = 'provider';
    const providerPathTs = path.join(__dirname, `${providerFileName}.ts`);
    const providerPathJs = path.join(__dirname, `${providerFileName}.js`);

    const onEventHandler = new lambdanodejs.NodejsFunction(
      this,
      `${this.id}-OnEventHandler`,
      {
        entry: fs.existsSync(providerPathTs) ? providerPathTs : providerPathJs,
        handler: 'onEvent',
        runtime: lambda.Runtime.NODEJS_18_X,
        timeout: cdk.Duration.minutes(5),
      },
    );

    onEventHandler.addToRolePolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        resources: [this.props.user.userArn],
        actions: [
          'iam:CreateAccessKey',
          'iam:DeleteAccessKey',
        ],
      }),
    );

    onEventHandler.addToRolePolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        resources: [
          `arn:aws:ssm:*:*:parameter${this.props.parameterName}`, // Parameter name starts with a slash
        ],
        actions: [
          'ssm:DeleteParameter',
          'ssm:DescribeParameters',
          'ssm:GetParameter',
          'ssm:GetParameterHistory',
          'ssm:PutParameter',
        ],
      }),
    );

    return new cr.Provider(this, `${this.id}-CreateAccessKeyProvider`, {
      onEventHandler,
    });
  }

  private createResource(): cdk.CustomResource {
    return new cdk.CustomResource(this, `${this.id}-Resource`, {
      serviceToken: this.provider.serviceToken,
      resourceType: 'Custom::SSMSecureIAMAccessKey',
      properties: {
        parameterName: this.props.parameterName,
        region: this.props.region,
        userName: this.props.user.userName,
      },
    });
  }
}
