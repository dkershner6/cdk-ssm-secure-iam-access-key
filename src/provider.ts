import {
  IAMClient,
  CreateAccessKeyCommand,
  DeleteAccessKeyCommand,
} from '@aws-sdk/client-iam';
import {
  SSMClient,
  DeleteParameterCommand,
  PutParameterCommand,
} from '@aws-sdk/client-ssm';
import type {
  CloudFormationCustomResourceCreateEvent,
  CloudFormationCustomResourceDeleteEvent,
  CloudFormationCustomResourceEvent,
  CloudFormationCustomResourceResponse,
  CloudFormationCustomResourceUpdateEvent,
} from 'aws-lambda';

/** The IAMAccessKey created is stored JSON.stringified using this interface */
export interface IAMAccessKey {
  accessKeyId: string;
  secretAccessKey: string;
}

const getParameterName = (
  resourceProperties: CloudFormationCustomResourceCreateEvent['ResourceProperties'],
): string => {
  const parameterName = resourceProperties?.parameterName;
  if (!parameterName) {
    throw new Error('parameterName is not set');
  }
  return parameterName;
};

const getRegion = (
  resourceProperties: CloudFormationCustomResourceCreateEvent['ResourceProperties'],
): string => {
  const region = resourceProperties?.region ?? process.env.AWS_REGION;
  if (!region) {
    throw new Error('region or AWS_REGION is not set');
  }
  return region;
};

const getUserUserName = (
  resourceProperties: CloudFormationCustomResourceCreateEvent['ResourceProperties'],
): string => {
  const userName = resourceProperties?.userName;
  if (!userName) {
    throw new Error('userName is not set');
  }
  return userName;
};

export const onEvent = async (
  event: CloudFormationCustomResourceEvent,
): Promise<CloudFormationCustomResourceResponse> => {
  console.debug('Event Received:', JSON.stringify(event));

  try {
    switch (event.RequestType) {
      case 'Create':
        return await onCreate(
          event as CloudFormationCustomResourceCreateEvent,
        );
      case 'Update':
        return await onUpdate(
          event as CloudFormationCustomResourceUpdateEvent,
        );
      case 'Delete':
        return await onDelete(
          event as CloudFormationCustomResourceDeleteEvent,
        );
      default:
        throw new Error(`Unknown event type in event ${event}`);
    }
  } catch (err) {
    console.error(err);
    throw new Error('Failed');
  }
};

const onCreate = async (
  event: CloudFormationCustomResourceCreateEvent,
): Promise<CloudFormationCustomResourceResponse> => {
  const region = getRegion(event.ResourceProperties);
  const userName = getUserUserName(event.ResourceProperties);

  const iam = new IAMClient({ region });
  const accessKeyResponse = await iam.send(
    new CreateAccessKeyCommand({ UserName: userName }),
  );

  if (!accessKeyResponse?.AccessKey?.AccessKeyId) {
    throw new Error('AccessKey not created');
  }

  console.debug(
    'AccessKey created:',
    accessKeyResponse?.AccessKey?.AccessKeyId,
  );

  const parameterName = getParameterName(event.ResourceProperties);

  const ssm = new SSMClient({ region });

  await ssm.send(
    new PutParameterCommand({
      Name: parameterName,
      Type: 'SecureString',
      Value: JSON.stringify({
        accessKeyId: accessKeyResponse.AccessKey.AccessKeyId,
        secretAccessKey: accessKeyResponse?.AccessKey.SecretAccessKey,
      } as IAMAccessKey),
      Overwrite: true,
    }),
  );

  console.debug('Parameter created:', parameterName);

  return {
    Status: 'SUCCESS',
    RequestId: event.RequestId,
    StackId: event.StackId,
    LogicalResourceId: event.LogicalResourceId,
    PhysicalResourceId: accessKeyResponse.AccessKey.AccessKeyId,
  };
};

const onUpdate = async (
  event: CloudFormationCustomResourceUpdateEvent,
): Promise<CloudFormationCustomResourceResponse> => {
  return {
    Status: 'SUCCESS',
    RequestId: event.RequestId,
    StackId: event.StackId,
    LogicalResourceId: event.LogicalResourceId,
    PhysicalResourceId: event.PhysicalResourceId,
  };
};

const onDelete = async (
  event: CloudFormationCustomResourceDeleteEvent,
): Promise<CloudFormationCustomResourceResponse> => {
  const parameterName = getParameterName(event.ResourceProperties);
  const region = getRegion(event.ResourceProperties);
  const userName = getUserUserName(event.ResourceProperties);
  const accessKeyId = event.PhysicalResourceId;

  const ssm = new SSMClient({ region });

  await ssm.send(
    new DeleteParameterCommand({
      Name: parameterName,
    }),
  );

  console.debug('Parameter deleted:', parameterName);

  const iam = new IAMClient({ region });

  await iam.send(
    new DeleteAccessKeyCommand({
      UserName: userName,
      AccessKeyId: accessKeyId,
    }),
  );

  console.debug('AccessKey deleted:', { userName, accessKeyId });

  return {
    Status: 'SUCCESS',
    RequestId: event.RequestId,
    StackId: event.StackId,
    LogicalResourceId: event.LogicalResourceId,
    PhysicalResourceId: event.PhysicalResourceId,
  };
};

