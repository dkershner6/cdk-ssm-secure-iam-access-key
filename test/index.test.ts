import { assertions, App, Stack } from 'aws-cdk-lib';
import { ISSMSecureIAMAccessKeyProps, SSMSecureIAMAccessKey } from '../src';

const mockApp = new App();
const stack = new Stack(mockApp, 'TestStack');

it('Should deploy', () => {
  const testProps: ISSMSecureIAMAccessKeyProps = {
    parameterName: '/test/param',
    // @ts-expect-error - Incomplete Mock
    user: {
      userName: 'testUserName',
      userArn: 'testUserArn',

    },
  };

  new SSMSecureIAMAccessKey(stack, 'test', testProps);

  expect(assertions.Template.fromStack(stack).toJSON()).toMatchSnapshot();
});