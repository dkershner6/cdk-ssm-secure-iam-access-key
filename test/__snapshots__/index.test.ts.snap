// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should deploy 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "testtestCreateAccessKeyProviderframeworkonEventA478602C": {
      "DependsOn": [
        "testtestCreateAccessKeyProviderframeworkonEventServiceRoleDefaultPolicy8B69BC01",
        "testtestCreateAccessKeyProviderframeworkonEventServiceRoleD0A9904F",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "7382a0addb9f34974a1ea6c6c9b063882af874828f366f5c93b2b7b64db15c94.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (TestStack/test/test-CreateAccessKeyProvider)",
        "Environment": {
          "Variables": {
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "testtestOnEventHandlerCCBBF9B1",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": {
          "Fn::GetAtt": [
            "testtestCreateAccessKeyProviderframeworkonEventServiceRoleD0A9904F",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "testtestCreateAccessKeyProviderframeworkonEventServiceRoleD0A9904F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "testtestCreateAccessKeyProviderframeworkonEventServiceRoleDefaultPolicy8B69BC01": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "testtestOnEventHandlerCCBBF9B1",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "testtestOnEventHandlerCCBBF9B1",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testtestCreateAccessKeyProviderframeworkonEventServiceRoleDefaultPolicy8B69BC01",
        "Roles": [
          {
            "Ref": "testtestCreateAccessKeyProviderframeworkonEventServiceRoleD0A9904F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testtestOnEventHandlerCCBBF9B1": {
      "DependsOn": [
        "testtestOnEventHandlerServiceRoleDefaultPolicy7304DCEC",
        "testtestOnEventHandlerServiceRoleD0EFD869",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "1588857bb0ca3aec27fdd0ef37a2dd1743332db03479ad96c0edbc865659bc39.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.onEvent",
        "Role": {
          "Fn::GetAtt": [
            "testtestOnEventHandlerServiceRoleD0EFD869",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "testtestOnEventHandlerServiceRoleD0EFD869": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "testtestOnEventHandlerServiceRoleDefaultPolicy7304DCEC": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "iam:CreateAccessKey",
                "iam:DeleteAccessKey",
              ],
              "Effect": "Allow",
              "Resource": "testUserArn",
            },
            {
              "Action": [
                "ssm:DeleteParameter",
                "ssm:DescribeParameters",
                "ssm:GetParameter",
                "ssm:GetParameterHistory",
                "ssm:PutParameter",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:ssm:*:*:parameter/test/param",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testtestOnEventHandlerServiceRoleDefaultPolicy7304DCEC",
        "Roles": [
          {
            "Ref": "testtestOnEventHandlerServiceRoleD0EFD869",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testtestResourceDC276495": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "testtestCreateAccessKeyProviderframeworkonEventA478602C",
            "Arn",
          ],
        },
        "parameterName": "/test/param",
        "userName": "testUserName",
      },
      "Type": "Custom::SSMSecureIAMAccessKey",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
